### PR TITLE
Add QSD vhost-user-blk-pci device options test

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2154781
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=qsd_vubp_options --iothread_scheme=roundrobin --nr_iothreads=2 --platform=x86_64 --guestname=RHEL.9.2.0,Win2019 --driveformat=virtio_scsi --nicmodel=virtio_net --imageformat=qcow2 --machines=q35,i440fx   --firmware=default_bios --netdst=virbr0 --clone=no --customsparams="cpu_model_flags = 'vmx,+kvm_pv_unhalt'"
 (01/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1: STARTED
 (01/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1: PASS (45.33 s)
 (02/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1.q35: STARTED
 (02/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1.q35: PASS (45.69 s)
 (03/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2: STARTED
 (03/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2: PASS (45.32 s)
 (04/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2.q35: STARTED
 (04/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2.q35: PASS (45.95 s)
 (05/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed: STARTED
 (05/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed: PASS (46.20 s)
 (06/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: STARTED
 (06/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: PASS (46.76 s)
 (07/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1: STARTED
 (07/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1: PASS (97.54 s)
 (08/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1.q35: STARTED
 (08/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_1.q35: PASS (97.14 s)
 (09/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2: STARTED
 (09/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2: PASS (96.73 s)
 (10/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2.q35: STARTED
 (10/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_2.q35: PASS (94.85 s)
 (11/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed: STARTED
 (11/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed: PASS (96.87 s)
 (12/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: STARTED
 (12/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: PASS (97.12 s)

File: qemu/tests/qsd_vubp_options.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                if attr[1] == disk_size and attr[2] == "disk":
+                    disk = kname
+                    break
+        else:
+            disks = utils_disk.get_windows_disks_index(session, disk_size)
+            disk = disks[0] if disks else None
+
+        exist_flag = True if disk else False
+        if check_exist_flag is not None:
+            if exist_flag != check_exist_flag:
```
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/qsd_vubp_options.py]: udpated

File: qemu/tests/cfg/qsd_vubp_options.cfg [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    qsd_image_export_stg1 = {"type":"vhost-user-blk","num-queues":1}
+    drive_format_stg1 = vhost-user-blk-pci
+    Linux:
+        guest_cmd = "dd if=/dev/urandom of=%s oflag=direct bs=1M count=1000"
+    Windows:
+        guest_cmd = "WIN_UTILS:\coreutils\DummyCMD.exe %s 1024000000 1"
+    ovmf:
+        no i440fx
+    variants:
+      - with_mix_1:
```
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/cfg/qsd_vubp_options.cfg]: change to with_mix_aim_off it means ats+indirect+multifunction off

File: qemu/tests/qsd_vubp_options.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    try:
+        qsd_name = params["qsd_namespaces"]
+        qsd = QsdDaemonDev(qsd_name, params)
+        qsd.start_daemon()
+        img = params["qsd_images_%s" % qsd_name]
+        add_vubp_into_boot(img, params, 6)
+
+        params["start_vm"] = 'yes'
+
+        login_timeout = int(params.get("login_timeout", 180))
```
Comment (menli820) [inline qemu/tests/qsd_vubp_options.py:72]:
```suggestion
       login_timeout = params.get_numeric("login_timeout", 180)
```

Comment (qingwangrh) [reply qemu/tests/qsd_vubp_options.py]: updated

**Comment by menli820** [comment]
LGTM

 (1/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_off.q35: STARTED
 (1/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_off.q35: PASS (131.75 s)
 (2/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_on.q35: STARTED
 (2/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_on.q35: PASS (147.50 s)
 (3/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: STARTED
 (3/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: PASS (149.65 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by menli820** [comment]
LGTM

 (1/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_off.q35: STARTED
 (1/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_off.q35: PASS (131.75 s)
 (2/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_on.q35: STARTED
 (2/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_on.q35: PASS (147.50 s)
 (3/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: STARTED
 (3/3) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: PASS (149.65 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Review by YongxueHong — APPROVED** [review]
LGTM.

File: qemu/tests/qsd_vubp_options.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
+"""QSD vhost-user-blk-pci device options test"""
+
+from provider.qsd import QsdDaemonDev, add_vubp_into_boot
+from virttest import error_context, env_process, utils_disk, utils_misc
+
+
+@error_context.context_aware
```
Comment (XueqiangWei) [inline qemu/tests/qsd_vubp_options.py:7]: no error_context.context() statement in codes, I think we can remove it. Others LGTM.

 (1/6) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_off.q35: PASS (75.67 s)
 (2/6) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_on.q35: PASS (73.56 s)
 (3/6) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: PASS (72.98 s)
 (4/6) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_off.q35: PASS (164.98 s)
 (5/6) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_mix_aim_on.q35: PASS (166.20 s)
 (6/6) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.qsd_vubp_options.with_packed.q35: PASS (161.65 s)

Comment (qingwangrh) [reply qemu/tests/qsd_vubp_options.py]: updated

**Review by YongxueHong — APPROVED** [review]

**Review by XueqiangWei — APPROVED** [review]

**Pull request merged**